### PR TITLE
First complete header generator version

### PR DESCRIPTION
**Pull request by Equidem**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,127 @@
...
+
+- [Installation](#installation)
+- [Usage](#usage)
+- [Examples](#examples)
+- [API Reference](#api-reference)
+
+<!-- tocstop -->
+
+## Installation
+Add this repository as an npm dependency and run the npm install command. No further setup is needed after that.
```
Comment (petrpatek) [inline README.md]: Maybe we can add a command like this? 
```
npm install header-generator
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,127 @@
...
+let headers = headersGenerator.getHeaders({
+        operatingSystems: [
+            "linux"
+        ],
+        locales: ["en-US", "en"]
+});
+```
+This method always generates a random realistic set of headers, excluding the request dependant headers, which need to be filled in afterwards. Since the generation is randomized, multiple calls to this method with the same parameters can generate multiple different outputs.
+## Examples
+A result that can be generated for the usage example above:
```
Comment (petrpatek) [inline README.md:43]: Maybe just "## Example result" ?

File: package.json [hunk 3]
```diff
@@ -1,23 +1,63 @@
 {
-  "name": "headers-generator",
-  "version": "0.0.1",
-  "description": "NodeJs package for generating browser-like headers.",
-  "main": "main.js",
-  "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
-  },
-  "repository": {
+  "_args": [
```
Comment (petrpatek) [inline package.json]: Could you please explain why is this here?

File: src/browserHelperFile.json [hunk 4]
```diff
@@ -0,0 +1 @@
...1|2","firefox/48.0|2","chrome/88.0.4324.109|2","chrome/73.0.3683.75|2","chrome/48.0.2564.97|2","chrome/89.0.4389.76|2","chrome/72.0.3626.121|2","chrome/91.0.4440.0|2","chrome/84.0.4147.110|2","firefox/71.0|2","chrome/90.0.4405.0|2","firefox/57.0|2","chrome/67.0.3396.87|2","chrome/88.0.4324.153|2","firefox/77.0|2","safari/604.4.7|2","chrome/80.0.3987.122|1","firefox/56.0|1","chrome/89.0.4389.86|2","firefox/75.0|2","chrome/88.0.4298.0|2","chrome/74.0.3729.172|2","chrome/87.0.4280.107|2","chrome/88.0.4308.0|2","chrome/88.0.4313.1|2","chrome/88.0.4324.79|2","chrome/88.0.4307.2|2","chrome/89.0.4371.0|2","chrome/87.0.4280.86|2","firefox/80.0|2","chrome/1.0.154.50|2","chrome/74.0.3729.157|2","chrome/70.0.3538.25|2","chrome/63.0.3239.84|2","chrome/80.0.3987.132|1","chrome/72.0.3626.81|2","chrome/88.8.383.407|2","chrome/86.0.4240.197|2","chrome/86.0.4240.75|1","chrome/87.0.4280.88|1","chrome/58.0.3029.83|2","firefox/64.0|2","firefox/78.0|1","chrome/89.0.4389.57|2","chrome/88.0.4288.0|2","chrome/11.0.696.71|2","firefox/3.6.10|2","chrome/88.0.4324.155|2","chrome/66.0.3359.158|2","firefox/46.0|2","chrome/48.0.2564.109|2","chrome/79.0.3945.116|2","chrome/90.0.4430.19|2","chrome/62.0.3202.94|2"]
```
Comment (petrpatek) [inline src/browserHelperFile.json]: I think we should stick for one file naming convention. Could you please use the `-` separator. Is the camelCase some json naming convention?

File: src/header-generator.js [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+};
+
+const http1SecFetchAttributes = {
+    mode: 'Sec-Fetch-Mode',
+    dest: 'Sec-Fetch-Dest',
+    site: 'Sec-Fetch-Site',
+    user: 'Sec-Fetch-User',
+};
+
+function getRandomInteger(minimum, maximum) {
```
Comment (petrpatek) [inline src/header-generator.js]: Maybe we can add some simple jsdoc even for our internal functions? It is super helpful for IDE completions and hints.
I think the param typing should do.

File: src/header-generator.js [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    return minimum + Math.floor(Math.random() * (maximum - minimum + 1));
+}
+
+function shuffleArray(array) {
+    if (array.length > 1) {
+        for (let x = 0; x < 10; x++) {
```
Comment (petrpatek) [inline src/header-generator.js]: Why it swaps always 10 items and not based on the number of items in the array? Just curious here 🧑‍🎓

File: src/header-generator.js [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    return {
+        name: nameVersionSplit[0],
+        version: preparedVersion,
+        completeString: browserString,
+    };
+}
+
+const headerGeneratorOptionsShape = {
+    browsers: ow.optional.array,
+    operatingSystems: ow.optional.array,
```
Comment (petrpatek) [inline src/header-generator.js]: I think we should validate also the contents of the array.

File: src/header-generator.js [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        const inputSample = this.inputGeneratorNetwork.generateSampleWheneverPossible(possibleAttributeValues);
+
+        if (!inputSample) {
+            throw new Error('No headers based on this input can be generated. Please relax or change some of the requirements you specified.');
+        }
+
+        // Generate the actual headers
+        const generatedSample = this.headerGeneratorNetwork.generateSample(inputSample);
+
+        // Manually fill the accept-language header with random ordering of the locales from input
```
Comment (petrpatek) [inline src/header-generator.js:218]: Maybe we can put the accept-language to a separate function?

File: src/header-generator.js [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        }
+
+        let acceptLanguageFieldValue = localesInAddingOrder[0];
+        for (let x = 1; x < localesInAddingOrder.length; x++) {
+            acceptLanguageFieldValue += `,${localesInAddingOrder[x]};${1 - x * 0.1}`;
+        }
+
+        generatedSample[acceptLanguageFieldName] = acceptLanguageFieldValue;
+
+        // Add fixed headers if needed
```
Comment (petrpatek) [inline src/header-generator.js:266]: Maybe we can move it to a separate function as well?

**Review by petrpatek — CHANGES_REQUESTED** [review]
Looks great! 🥇 I have just added some small comments and questions.

File: README.md (hunk 1 above)
Comment (Equidem) [reply README.md]: We would have to publish this as an npm package to do that, right?

File: package.json (hunk 3 above)
Comment (Equidem) [reply package.json]: Because I copied in the wrong package.json file and didn't notice. Will get rid of the ballast.

File: src/browserHelperFile.json (hunk 4 above)
Comment (Equidem) [reply src/browserHelperFile.json]: Used it to obviously distinguish the data files from the code files. But will use the - separator if you think it is better that way.

File: src/header-generator.js (hunk 5 above)
Comment (Equidem) [reply src/header-generator.js]: I generally don't like using comments for these short helper functions, since they should be self-explanatory. But will add the param typing for prepareHttpBrowserObject and prepareBrowserObject, since it might not be clear what the argument is supposed to be.

File: src/header-generator.js (hunk 6 above)
Comment (Equidem) [reply src/header-generator.js]: It was originally named differently and was (and still is) used only for locales, where there can only be 10 at most. Didn't remember to change this when I changed the name, thanks for the catch.

File: src/header-generator.js (hunk 7 above)
Comment (Equidem) [reply src/header-generator.js]: Wasn't sure how to do that without looping over the items and validating them one by one. Will probably do the looping option, let me know if you have any better idea.

File: src/header-generator.js (hunk 8 above)
Comment (Equidem) [reply src/header-generator.js:218]: Going from that logic, this method should be splintered to 4-5 functions. I prefer it this way, since a reader can read it all in uninterrupted progression without jumping around to different function, but I get that somebody might dislike it. Do you think it should be changed?

File: src/browserHelperFile.json (hunk 4 above)
Comment (petrpatek) [reply src/browserHelperFile.json]: I would place them in a subfolder and used the separator.

File: src/header-generator.js (hunk 5 above)
Comment (petrpatek) [reply src/header-generator.js]: Ok, make sense 🆗

File: src/header-generator.js (hunk 7 above)
Comment (petrpatek) [reply src/header-generator.js]: I think we can use this https://sindresorhus.com/ow/classes/arraypredicate.html#validate, or custom validator. There is an example from SDK, but there are no arrays. https://github.com/apify/apify-js/blob/9f94909103b801419ee6b367f85e5b698b7c2530/src/validators.js#L9

File: src/header-generator.js (hunk 8 above)
Comment (petrpatek) [reply src/header-generator.js:218]: That makes sense. I enjoyed reading it, but it was here when the chapter got kind of long. But this is definitely a matter of opinion. I think we can keep it this way.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,149 @@
...
+
+* * *
+
+<a name="prepareHttpBrowserObject"></a>
+
+### `prepareHttpBrowserObject(httpBrowserString)`
+
+| Param | Type | Description |
+| --- | --- | --- |
+| httpBrowserString | <code>string</code> | a string containing the browser name, version and http version, such as "chrome/88.0.4324.182|2" |
```
Comment (mnmkng) [inline README.md]: What does this do?

File: README.md [hunk 11]
```diff
@@ -1,2 +1,149 @@
...
+
+* * *
+
+<a name="prepareBrowserObject"></a>
+
+### `prepareBrowserObject(browserString)`
+
+| Param | Type | Description |
+| --- | --- | --- |
+| browserString | <code>string</code> | a string containing the browser name and version, such as "chrome/88.0.4324.182" |
```
Comment (mnmkng) [inline README.md]: Same here. Some examples or explanation would be nice.

File: README.md [hunk 12]
```diff
@@ -1,2 +1,149 @@
...
+
+
+* * *
+
+<a name="Browser"></a>
+
+### `Browser`
```
Comment (mnmkng) [inline README.md]: I would name this type `BrowserDescription` or `BrowserDefinition` or something like that. It's not really a `Browser`.

File: README.md [hunk 13]
```diff
@@ -1,2 +1,149 @@
...
+<a name="HeaderGeneratorOptions"></a>
+
+### `HeaderGeneratorOptions`
+
+| Param | Type | Description |
+| --- | --- | --- |
+| browsers | [<code>Array.&lt;Browser&gt;</code>](#Browser) | List of Browsers to generate the headers for. |
+| operatingSystems | <code>Array.&lt;string&gt;</code> | List of operating systems to generate the headers for.  The options are "windows", "macos", "linux", "android" and "ios". |
+| devices | <code>Array.&lt;string&gt;</code> | List of devices to generate the headers for. Options are "desktop" and "mobile". |
+| locales | <code>Array.&lt;string&gt;</code> | List of at most 10 languages to include in the `Accept-Language` request header. |
```
Comment (mnmkng) [inline README.md]: examples of the locales, their structure and how to find them would be great

Comment (mnmkng) [reply README.md]: I mean, "List of at most 10 languages" could also be [Czech, Slovak, English]

File: README.md [hunk 14]
```diff
@@ -1,2 +1,149 @@
...
+<a name="Browser"></a>
+
+### `Browser`
+
+| Param | Type | Description |
+| --- | --- | --- |
+| name | <code>string</code> | One of "chrome", "firefox" and "safari". |
+| minVersion | <code>number</code> | Minimal version of browser used. |
+| maxVersion | <code>number</code> | Maximal version of browser used. |
+| httpVersion | <code>string</code> | Either 1 or 2. If none specified the global `httpVersion` is used. |
```
Comment (mnmkng) [inline README.md]: Shouldn't the `httpVersion` be determined automatically depending on the version of the browser?

Comment (petrpatek) [reply README.md]: You can have http2 supported in the browser and not on the website, so you want to get HTTP1 headers even from HTT2 browser.

File: README.md (hunk 13 above)
Comment (petrpatek) [reply README.md]: I guess we can add smth like - "list of 10 Accept-Language compatible languages" and add link to the men https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language

File: README.md (hunk 10 above)
Comment (petrpatek) [reply README.md:105]: I guess this should not be in public docs, right?

File: README.md [hunk 15]
```diff
@@ -1,2 +1,149 @@
...
+| Param | Type | Description |
+| --- | --- | --- |
+| options | [<code>HeaderGeneratorOptions</code>](#HeaderGeneratorOptions) | specifies options that should be overridden for this one call |
+
+
+* * *
+
+<a name="prepareHttpBrowserObject"></a>
+
+### `prepareHttpBrowserObject(httpBrowserString)`
```
Comment (petrpatek) [inline README.md]: I guess this should not be part of the public docs, right?

**Review by petrpatek — CHANGES_REQUESTED** [review]
Looks good! Just a small documentation thingy.

File: src/header-generator.js [hunk 16]
```diff
@@ -0,0 +1,284 @@
+const fs = require('fs');
+const path = require('path');
+const { BayesianNetwork } = require('bayesian-network');
+const { default: ow } = require('ow');
+
+const headerNetworkDefinitionPath = path.join(__dirname, './data_files/header-network-definition.json');
+const inputNetworkDefinitionPath = path.join(__dirname, './data_files/input-network-definition.json');
+const browserHelperFilePath = path.join(__dirname, './data_files/browser-helper-file.json');
```
Comment (sestys) [inline src/header-generator.js]: What exactly should this file be? The pre-trained models?

File: README.md (hunk 14 above)
Comment (mnmkng) [reply README.md]: Yeah, my bad. I was referring to the `If none specified the global `httpVersion` is used.` part. I would expect it to do this: `If none specified, it will use the headers which the browser uses by default`

File: src/header-generator.js (hunk 16 above)
Comment (Equidem) [reply src/header-generator.js]: Yeah. There will be two more repositories with pull requests in the next few days, first one is the bayesian network, second is the one that creates these files from data.

File: README.md (hunk 14 above)
Comment (Equidem) [reply README.md]: @petrpatek @mnmkng, so how should this work? The interface is according to what Petr specified, but it can be changed, of course.

File: README.md (hunk 13 above)
Comment (Equidem) [reply README.md]: Will do.

File: README.md (hunk 14 above)
Comment (mnmkng) [reply README.md]: Well, I don't know how it works, but I would expect it to work this way:

- no `httpVersion` configuration, default for browser is used.
- if local `httpVersion` is specified, it will be used.
- if global `httpVersion` is specified and local is not specified, that one will be used.

Importantly, there must not be any "default global", "default local" `httpVersion` which would override the natural HTTP version of the browser whose headers get generated. It can only be overridden explicitly by the user.

If it works like this, then it's fine, but the `Either 1 or 2. If none specified the global `httpVersion` is used.` does not explain this very well.

Comment (Equidem) [reply README.md]: Currently, there is a default global httpVersion of 2. Looking at it, I forgot to include the comment that mentioned it.
Anyway, it should be extremely easy to get rid of that default, so I will do that.

File: README.md (hunk 15 above)
Comment (Equidem) [reply README.md]: Definitely not, thanks for the catch.

File: README.md (hunk 14 above)
Comment (petrpatek) [reply README.md]: @mnmkng I get your point. The problem is that the HTTP version is based on the website, not the browser. There is no default for the browser, but only default for the website. HTTP 2 is the default value because most modern websites are using http2. We cannot determine the HTTP version of the target website in this package. This will make the dev open the network activity and see the actual requests that the browser does. We could add this auto HTTP resolving to the got-scraping package - it should be easy with the HTTP resolver we already have in place.

Comment (mnmkng) [reply README.md]: Yeah, just be careful. Getting rid of the default does not automatically mean that the natural browser version will be used. I don't know how that's implemented.

Comment (Equidem) [reply README.md]: @petrpatek Ok, so I will leave the default and just change the comments a little so that it is clear what version will be used in which case. Please, let me know if you and Ondra decide to change it after all.

File: src/data_files/headers-order.json [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        "Sec-Fetch-User",
+        "accept-encoding",
+        "Accept-Encoding",
+        "accept-language",
+        "Accept-Language",
+        "dnt",
+        "DNT",
+        "te"
+    ],
+    "firefox": [
```
Comment (mnmkng) [inline src/data_files/headers-order.json:62]: In Firefox, when I show Raw headers, they are in this order, which is different from the order shown here.

```
GET /apify/header-generator/pull/6/files/83a049bad3c8ac6c912c6aea2fc6bf1be436d104..5871a061e28123d2eeba8c9737d20f565f687e04 HTTP/2
Host: github.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:87.0) Gecko/20100101 Firefox/87.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Referer: https://github.com/notifications?query=
DNT: 1
Connection: keep-alive
Cookie: 'some-cookie'
Upgrade-Insecure-Requests: 1
TE: Trailers
```

Also, interestingly, they are Capitalized, even though the connection shows HTTP/2 🤷‍♂️ @petrpatek 

image

Comment (Equidem) [reply src/data_files/headers-order.json:62]: The orderings are not final yet (I set it up only for one of the browsers so far and the others are just copies), that is why I have not re-requested a review. I will finish them later today.

Comment (Equidem) [reply src/data_files/headers-order.json:62]: The capitalized HTTP 2 is weird, I will check whether we have these cases in the headers we gathered as data.

Comment (Equidem) [reply src/data_files/headers-order.json:62]: I checked the data and out of the first 38k items we gathered, there is not a single HTTP 2 one that has the User-Agent header filled, all of them use the user-agent header as they should.

File: src/header-generator.js [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+/**
+ * HeaderGenerator randomly generates realistic browser headers based on specified options.
+ */
+class HeaderGenerator {
+    /**
+     * @param {HeaderGeneratorOptions} options - default header generation options used unless overridden
+     */
+    constructor(options = {}) {
+        ow(options, 'HeaderGeneratorOptions', ow.object.exactShape(headerGeneratorOptionsShape));
+        this.defaultOptions = options;
```
Comment (mnmkng) [inline src/header-generator.js]: Be careful with assignments like this. The object is passed by reference so if at any point later you change the `defaultOptions` you also mutate the `options`.

```js
// somewhere in your code
this.defaultOptions.foo = 'bar'


// somewhere in user's code
options.foo === 'bar' // true
```

You might know that you're not modifying the `defaultOptions` anywhere, but someone else, later, could. So the better way of doing this is to make a copy of the object.

File: src/header-generator.js [hunk 19]
```diff
@@ -0,0 +1,310 @@
...
+        this.headersOrder = JSON.parse(fs.readFileSync(headersOrderFilePath, { encoding: 'utf8' }));
+        const uniqueBrowserStrings = JSON.parse(fs.readFileSync(browserHelperFilePath, { encoding: 'utf8' }));
```
Comment (mnmkng) [inline src/header-generator.js]: Can the files change when the `HeaderGenerator` runs? Those are pretty expensive operations, so it would make sense to open those files at `require` time and not for each `new HeaderGenerator` call.

File: src/header-generator.js [hunk 20]
```diff
@@ -0,0 +1,310 @@
+const fs = require('fs');
+const path = require('path');
+const { BayesianNetwork } = require('bayesian-network');
+const { default: ow } = require('ow');
+
+const headerNetworkDefinitionPath = path.join(__dirname, './data_files/header-network-definition.json');
+const inputNetworkDefinitionPath = path.join(__dirname, './data_files/input-network-definition.json');
+const browserHelperFilePath = path.join(__dirname, './data_files/browser-helper-file.json');
+const headersOrderFilePath = path.join(__dirname, './data_files/headers-order.json');
```
Comment (mnmkng) [inline src/header-generator.js]: See comments on line 144 and 156. It would be better to open those files immediately and once, instead of doing it repeatedly in the constructor. You can simply `require` a JSON and you will get the parsed JSON.

```js
const headerNetworkDefinition = require('./data_files/header-network-definition.json'));
headerNetworkDefinition.foo = 'bar';
```

File: src/header-generator.js [hunk 21]
```diff
@@ -0,0 +1,310 @@
...
+            if (browserString === missingValueDatasetToken) {
+                this.uniqueBrowsers.push({
+                    name: missingValueDatasetToken,
+                });
+            } else {
+                this.uniqueBrowsers.push(prepareHttpBrowserObject(browserString));
+            }
+        }
+        this.inputGeneratorNetwork = new BayesianNetwork(inputNetworkDefinitionPath);
+        this.headerGeneratorNetwork = new BayesianNetwork(headerNetworkDefinitionPath);
```
Comment (mnmkng) [inline src/header-generator.js]: Same here. Does not make sense to load and parse the files repeatedly. We should use the loaded definition instead of the path, or we need to cache the path / object in the `BayesianNetwork` module. I would just use the loaded object as the parameter.

It's often cumbersome to work with constructors that accept paths, because if you want to programmatically change something in the input, you need to write the changes to a file instead of just modifying an object and be done with it.

File: src/header-generator.js [hunk 22]
```diff
@@ -0,0 +1,310 @@
...
+     * @param {HeaderGeneratorOptions} options - specifies options that should be overridden for this one call
+     */
+    getHeaders(options = {}) {
+        ow(options, 'HeaderGeneratorOptions', ow.object.exactShape(headerGeneratorOptionsShape));
+        const headerOptions = { ...this.defaultOptions, ...options };
+
+        // Set up defaults
+        if (!headerOptions.locales) {
+            headerOptions.locales = ['en-US'];
+        }
```
Comment (mnmkng) [inline src/header-generator.js]: Would it make sense to setup all those defaults in the constructor already? Or at least use a function `setDefaultOptions` or smth

File: src/header-generator.js [hunk 23]
```diff
@@ -0,0 +1,310 @@
...
+                'linux',
+                'android',
+                'ios',
+            ];
+        }
+
+        headerOptions.browsers = headerOptions.browsers.map((browserObject) => {
+            if (!browserObject.httpVersion) {
+                browserObject.httpVersion = headerOptions.httpVersion;
+            }
```
Comment (mnmkng) [inline src/header-generator.js:180]: Correct me if I'm wrong, but:

- If the user does not provide `headerOptions.httpVersion` it will be set to `2`.
- If the user does not provide `headerOptions.browsers` or provides them, but without `browserObject.httpVersion`, the `browserObject.httpVersion` will be set to `headerOptions.httpVersion` which will be `2`. 

Both of these situations will happen quite often I assume. Now, the user did not specify a browser version, yet, I assume, for browsers that we don't have any HTTP2 headers for, this will throw and tell the user to "relax" the requirements. The user can't relax anything though, because the problem is caused by the defaults.

File: src/header-generator.js [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+
+const headerNetworkDefinitionPath = path.join(__dirname, './data_files/header-network-definition.json');
+const inputNetworkDefinitionPath = path.join(__dirname, './data_files/input-network-definition.json');
+const browserHelperFilePath = path.join(__dirname, './data_files/browser-helper-file.json');
+const headersOrderFilePath = path.join(__dirname, './data_files/headers-order.json');
+
+const browserHttpNodeName = '*BROWSER_HTTP';
+const operatingSystemNodeName = '*OPERATING_SYSTEM';
+const deviceNodeName = '*DEVICE';
+const missingValueDatasetToken = '*MISSING_VALUE*';
```
Comment (mnmkng) [inline src/header-generator.js]: I would expect those to be in CAPS as they are top level constants. I was looking for those variables in scope of the functions where they're used.

Comment (mnmkng) [reply src/header-generator.js]: Btw, why do all of them except `*MISSING_VALUE*` have a single asterisk? Is that a typo?

File: src/header-generator.js [hunk 25]
```diff
@@ -0,0 +1,310 @@
...
+        ow(options, 'HeaderGeneratorOptions', ow.object.exactShape(headerGeneratorOptionsShape));
+        this.defaultOptions = options;
+        this.headersOrder = JSON.parse(fs.readFileSync(headersOrderFilePath, { encoding: 'utf8' }));
+        const uniqueBrowserStrings = JSON.parse(fs.readFileSync(browserHelperFilePath, { encoding: 'utf8' }));
+        this.uniqueBrowsers = [];
+        for (const browserString of uniqueBrowserStrings) {
+            if (browserString === missingValueDatasetToken) {
+                this.uniqueBrowsers.push({
+                    name: missingValueDatasetToken,
+                });
```
Comment (mnmkng) [inline src/header-generator.js]: Why push this empty value into the array? This would deserve a comment at least.

**Review by mnmkng — CHANGES_REQUESTED** [review]
I'll do the rest later, need to switch to something else now.

File: src/header-generator.js (hunk 19 above)
Comment (Equidem) [reply src/header-generator.js]: It's written like this mainly to allow easy setup of the generator with different files when instantiating it in the future, though it can of course be done efficiently with lazy loading the files or something like that. I don't see it as an issue either way, since we don't expect people to (usually) create more than one instance of the generator. It is trivial to change this if you feel strongly about it, of course.

File: src/header-generator.js (hunk 21 above)
Comment (Equidem) [reply src/header-generator.js]: What you are suggesting goes against the fundamental principle of encapsulation. The HeaderGenerator class has no business knowing anything about the definitions of the BayesianNetwork, or even worse, changing them.
However, as mentioned before, we can certainly preload them if you feel strongly about that part.

File: src/header-generator.js (hunk 19 above)
Comment (mnmkng) [reply src/header-generator.js]: Well, the first user of the generator already did what you would not expect 😅 https://github.com/apify/got-scraping/pull/1#discussion_r594995242

File: src/header-generator.js (hunk 21 above)
Comment (mnmkng) [reply src/header-generator.js]: I don't think there's a difference in encapsulation in those two formats:

```js
const foo = new Dog(pathToConfig);
const bar = new Cat(config);
```

Neither needs to know anything about the config. It just needs to provide it in the correct format. Both are totally fine ways of doing things. But we need to make sure they perform well.

We can either:

```js
const networkDefinition = require('./network-definition.json');
// ...
const foo = new BayesianNetwork(networkDefinition);
```

or we need to implement lazy loading in the `BayesianNetwork`. Both are good solutions. The lazy loading is probably a bit better, but also more time consuming.

Comment (Equidem) [reply src/header-generator.js]: I meant the "It's often cumbersome to work with constructors that accept paths, because if you want to programmatically change something in the input, you need to write the changes to a file instead of just modifying an object and be done with it." comment, not the options of loading. As I added afterwards, those are ok.

File: src/header-generator.js (hunk 19 above)
Comment (Equidem) [reply src/header-generator.js]: I actually asked Petr how he uses it before answering the comment. Guess I only got the most recent answer :D
Anyway, I will make the files only load once then.

File: src/header-generator.js (hunk 24 above)
Comment (Equidem) [reply src/header-generator.js]: As to the constants, I would have made them all caps, but I think I checked one of our repositories and top level constants were written there this way, which is why I did it like this. Will try to remember which one it was and will make this ALL_CAPS.
The one vs two asterisk difference is actually there precisely to show that the constants are different things. The upper three are node names in the underlying bayesian network, the fourth is just a token representing empty value. I have added an empty line between them to reinforce the point.

File: src/header-generator.js (hunk 23 above)
Comment (Equidem) [reply src/header-generator.js:180]: That is precisely what would happen. As to the "they can't relax anything" point, another thing the error mentions is to change something in the requirements, which is what they would need to do here. Anyway, this is a logical consequence of enforcing the default of 2. If you think that it shouldn't work like this, maybe we can prefer HTTP 2 if the user doesn't specify he wants HTTP 1, but allow the generator to generate HTTP 1 headers if no HTTP 2 headers can be generated. Would that be better for the users? What does @petrpatek think?

File: src/header-generator.js (hunk 21 above)
Comment (mnmkng) [reply src/header-generator.js]: Yeah, I should've just left that second part of the comment out. It was only confusing.

File: src/header-generator.js [hunk 26]
```diff
@@ -0,0 +1,313 @@
...
+            const position1 = getRandomInteger(0, array.length - 1);
+            const position2 = getRandomInteger(0, array.length - 1);
+            const holder = array[position1];
+            array[position1] = array[position2];
+            array[position2] = holder;
+        }
+    }
+
+    return array;
+}
```
Comment (fnesveda) [inline src/header-generator.js]: This isn't a very good shuffling algorithm, it will leave a lot (~14%, see https://math.stackexchange.com/a/72351 for reasoning) of items unshuffled, as they will not get selected for the shuffling. Perhaps you could replace it with a better one, or maybe with `_.shuffle` from `underscore`?

**Review by fnesveda — CHANGES_REQUESTED** [review]
I've gone through the code, mostly through the BayesianNetwork logic, but since that package is not documented at all, it was a bit of reverse engineering, and I ran out of energy midway through, so I didn't review it super thoroughly. Anyway, I think the maths behind it is sound, so I think it should work.
I think it would really help future generations if the `bayesian-network` package and the process for generating the JSON data files was documented somewhere, it's pretty hard to understand what's going on without it.
Otherwise I just had one small comment in the rest of the code regarding the array shuffling, I didn't see any other issues not reported by the others.

Nice job, I hope it will work well! 😃  Did you wish you could write this in Python? 😄 I imagine some things would be a lot easier there.

File: src/header-generator.js (hunk 26 above)
Comment (Equidem) [reply src/header-generator.js]: I actually have the BayesianNetwork repository with absolute majority of the documentation done, I just needed to finish one thing before pushing it and creating a pull request for it too, but had my attention redirected elsewhere in the meantime. I had no idea someone would go through that code in this pull request, sorry :D 

I know the shuffling algorithm isn't great. It's nice that someone took the time to calculate how bad it is :D 
Anyway, it doesn't need to be great, I just use it to shuffle the locales a little so that they differ between various calls to getHeaders and I didn't want to add more dependencies needlessly.

As to whether I would rather use Python - I was actually toying with the idea of using it and making the generator a standalone script which would ran alongside the actor and communicate with him through message passing, but decided that this way it would be easier to use and that I don't actually need most of the advantages Python provides for it. Moreover, the repository that generates the definition files uses [DanfoJS](https://danfo.jsdata.org/), which is basically a JS copy of Pandas. The only problem is that it is still a little unfinished, so some things need to be worked around, but it was a great help anyway.

File: src/header-generator.js (hunk 23 above)
Comment (petrpatek) [reply src/header-generator.js:180]: I think this is a theoretical problem. I think we support only browsers we have data for - `chrome`, `firefox`, `safari` for http2. So the proposed scenario should never happen. If this happens, we would lack the support of the particular browser. 

We cant substitute HTTP2 headers with HTTP1 headers since it would fail the request because of incompatible headers such as `Connection`. 

If the user would specify more than one browser, let's say FF and safari. Safari would lack HTTP2 headers. The header generator would still produce HTTP2 headers for FF and ignore safari, am I right? Following this logic, the relaxing of requirements, in this case, means changing the browser, because there are no data for it. Once again this should never happen with default HTPP2.

Comment (Equidem) [reply src/header-generator.js:180]: It can happen for some browser versions, I guess - if someone picks a version of a browser (and only that one version) old enough that there are no HTTP 2 items in our data for that browser version.
You are right about the two-browsers case.
Anyway, @petrpatek and @mnmkng, please, discuss how this should work and let me know what you decide is best.

Comment (mnmkng) [reply src/header-generator.js:180]: If @petrpatek says it's ok then it's ok. I just wanted to know that this is not something that slipped through the cracks.

File: src/header-generator.js (hunk 26 above)
Comment (fnesveda) [reply src/header-generator.js]: Cool, I'll take a look at the commented BayesianNetwork version when it lands :) I was reading it in order to understand the logic of what's happening here, how does the generating work and what does it return and such.

As for the shuffling algorithm, I know it's not necessary to have it perfect in this case, there will usually probably be <10 locales in one array anyway, I'm just totally sure that one day in the future, when expanding this project, someone will need an array shuffling function for something else, will see that one is already implemented here, and will just use it without thinking twice. Anyway, how about something like this, adapted from Durstenfeld's array shuffling algorithm? It was included in Donald Knuth's The Art of Computer Programming, so I think it's reasonably good.

```javascript
function shuffleArray(array) {
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
}
```

**Comment by mnmkng** [comment]
I'll re-review after the tests pass.

**Comment by Equidem** [comment]
> I'll re-review after the tests pass.

They would have passed, but I first pushed changes here and then in the dependency, making them fail here. Sorry :)

File: src/header-generator.js (hunk 26 above)
Comment (mnmkng) [reply src/header-generator.js]: @fnesveda You actually read Donald Knuth's Art of Programming? 🙇

Comment (fnesveda) [reply src/header-generator.js]: @mnmkng Nah, I'm not that hardcore, but it was referenced to us a lot at school as "the algorithms bible" by many professors, and it kinda stuck, so I kinda implicitely trust stuff that comes from there :smile:

Comment (Equidem) [reply src/header-generator.js]: @fnesveda Sure, will definitely request your review on that one :)
Good point with others wanting to use that function not knowing it is not the best possible algorithm. The algorithm you suggested looks good, so I used it.

**Comment by mnmkng** [comment]
I think it's in a reasonable state now. I would still make some changes, like refactor the `getHeaders` function not to have 100 lines of code or add more tests, but I think it's mergeable for now and we can deal with those things later when we upgrade the package further.

**Review by mnmkng — APPROVED** [review]
